File: 그래프/[13023번] ABCDE/[13023번] ABCDE.py
# https://www.acmicpc.net/problem/13023

import sys

def dfs(p, c):
	visited[p] = True
	c += 1
	if c == 5:
		global flag
		flag = 1
		return
	for i in rel[p]:
		if visited[i] == False:
			dfs(i, c)
			if flag == 1:
				return
			visited[i] = False

N, M = map(int, sys.stdin.readline().split())
rel = [[] for _ in range(N)]
visited = [False for _ in range(N)]
flag = 0

for _ in range(M):
	s, d = map(int, sys.stdin.readline().split())
	rel[s].append(d)
	rel[d].append(s)	

for i in range(N):
	dfs(i, 0)
	visited = [False for _ in range(N)]
	if flag == 1:
		print("1\n")
		break
if flag == 0:
	print("0\n")

File: 그래프/[13023번] ABCDE/[13023번] ABCDE.cpp
// https://www.acmicpc.net/problem/13023

#include <bits/stdc++.h>

using namespace std;

int N, M;
vector<int>* rel;
int* visited;
int flag = 0;

void dfs(int p, int c){
	visited[p] = 1;
	c++;
	if (c == 5) {
		flag = 1;
		return;
	}
	for (int i = 0; i < rel[p].size(); i++){
		int next = rel[p][i];
		if (visited[next] == 0) {
			dfs(next, c);
			if (flag == 1)
				return;
			visited[next] = 0;
		}
	}
}

int main(void) {
	scanf("%d %d", &N, &M);
	
	rel = new vector<int>[N];
	visited = new int[N];
	fill_n(visited, N, 0);
	
	int s, d;
	for (int i = 0; i < M; i++) {
		scanf("%d %d", &s, &d);
		rel[s].push_back(d);
		rel[d].push_back(s);
	}
	
	for (int i = 0; i < N; i++) {
		dfs(i, 0);
		if (flag == 1){
			printf("1\n");
			break;
		}
		fill_n(visited, N, 0);
	}
	if (flag != 1)
		printf("0\n");
}